server/src/server.ts

```diff
@@ -1,314 +1,342 @@
-		if (!this.socket.writable)
-			return this.debug('Socket closed, dropping', pkt.type)
-		if (doCrypto && !this.cryptoPromise)
-			throw new Error(`Can't encrypt: handshake not finished`)
-
-		const writer = new BufWriter() // TODO size hint
-		writer.writeUInt32(0) // set later, but reserve space in buffer
-		encodePacket(pkt, writer)
-		let buf = writer.getBuffer()
-		buf.writeUInt32BE(buf.length - 4, 0) // write into space reserved above
-
+	private async INTERNAL_send(packet: ServerPacket, doCrypto: boolean = false) {
+		if (!this.socket.writable) {
+			this.debug("Socket closed, dropping", packet.type);
+			return;
+		}
+		if (doCrypto && !this.cryptoPromise) {
+			throw new Error("Can't encrypt: handshake not finished");
+		}
+		const packetWriter = new BufWriter(); // TODO: size hint (Ori: presumably means reducing unnecessary expansions)
+		packetWriter.writeUInt32(0); // set later, but reserve space in buffer
+		encodePacket(packet, packetWriter);
+		let packetBuffer = packetWriter.getBuffer();
+		packetBuffer.writeUInt32BE(packetBuffer.length - 4, 0); // write into space reserved above
 		if (doCrypto) {
-			const { cipher } = await this.cryptoPromise!
-			buf = cipher!.update(buf)
+			const { cipher } = await this.cryptoPromise!;
+			packetBuffer = cipher!.update(packetBuffer);
 		}
+		this.socket.write(packetBuffer);
+	}
 
-		this.socket.write(buf)
+	private async handlePacketReceived(packet: ClientPacket) {
+		if (!this.uuid) {
+			// not authenticated yet
+			switch (packet.type) {
+				case "Handshake":
+					await this.handleHandshakePacket(packet);
+					break;
+				case "EncryptionResponse":
+					await this.handleEncryptionResponsePacket(packet);
+					break;
+				default:
+					throw new Error(`Packet ${packet.type} from unauthenticated client ${this.id}`);
+			}
+		}
+		else {
+			await this.handler.handleClientPacketReceived(this, packet);
+		}
 	}
 
 	private async handleHandshakePacket(packet: HandshakePacket) {
-		if (this.cryptoPromise) throw new Error(`Already authenticated`)
-		if (this.verifyToken) throw new Error(`Encryption already started`)
-
-		this.modVersion = packet.modVersion
-		this.gameAddress = packet.gameAddress
-		this.claimedMojangName = packet.mojangName
-		this.world = packet.world
-		this.verifyToken = crypto.randomBytes(4)
-
-		await this.sendInternal({
-			type: 'EncryptionRequest',
-			publicKey: this.server.publicKeyBuffer,
-			verifyToken: this.verifyToken,
-		})
+		if (this.cryptoPromise) {
+			throw new Error("Already authenticated");
+		}
+		if (this.verifyToken) {
+			throw new Error("Encryption already started");
+		}
+		this.modVersion = packet.modVersion;
+		this.gameAddress = packet.gameAddress;
+		this.claimedMojangName = packet.mojangName;
+		this.world = packet.world;
+		this.verifyToken = crypto.randomBytes(4);
+		await this.INTERNAL_send({
+			type: "EncryptionRequest",
+			publicKey: PUBLIC_KEY_BUFFER,
+			verifyToken: this.verifyToken
+		});
```

Comment:
Should probably mention that this can't use the normal `send` method. Maybe its just me, but it was my first instict to think "why use the internal method?"